In [2]:
import pandas as pd
import numpy as np
import json

Importing the data as a dataframe.

In [34]:
original_dataset = pd.read_csv("data/Data_Lv3_UMDOTS_Escooters.csv")

Visualizing a few data point and checking the size of the dataset.

In [35]:
original_dataset.head()

Unnamed: 0  Unnamed: 0.1                    CREATED  \
0           0             0  2020-10-03 03:32:15+00:00   
1           1             1  2020-10-03 03:07:39+00:00   
2           2             2  2020-10-03 03:17:05+00:00   
3           3             3  2020-10-03 02:24:58+00:00   
4           4             4  2020-10-03 03:19:34+00:00   

                       START                        END  DISTANCE  RIDE_ID  \
0  2020-10-03 03:32:15+00:00  2020-10-03 03:32:46+00:00     0.008  3135665   
1  2020-10-03 03:07:39+00:00  2020-10-03 03:37:06+00:00     1.062  3135395   
2  2020-10-03 03:17:05+00:00  2020-10-03 03:24:24+00:00     0.847  3135503   
3  2020-10-03 02:24:58+00:00  2020-10-03 03:11:56+00:00     3.447  3134955   
4  2020-10-03 03:19:34+00:00  2020-10-03 03:22:45+00:00     0.386  3135529   

   CUSTOMER_ID  MINUTES VEHICLE_TYPE  \
0       193964        1      Scooter   
1       542947       30      Scooter   
2       517321        8      Scooter   
3       578051       47      Scooter   
4       552305        4      Scooter   

                                                PATH  \
0  {\n  "coordinates": [\n    [\n      -76.934163...   
1  {\n  "coordinates": [\n    [\n      -76.938180...   
2  {\n  "coordinates": [\n    [\n      -76.937599...   
3  {\n  "coordinates": [\n    [\n      -76.943553...   
4  {\n  "coordinates": [\n    [\n      -76.938115...   

                                          TIMESTAMPS  START LONG  START LAT  \
0  [\n  "2020-10-03T03:32:14Z",\n  "2020-10-03T03...  -76.934163  38.990656   
1  [\n  "2020-10-03T03:07:46Z",\n  "2020-10-03T03...  -76.938181  38.982087   
2  [\n  "2020-10-03T03:17:12Z",\n  "2020-10-03T03...  -76.937600  38.979613   
3  [\n  "2020-10-03T02:25:04Z",\n  "2020-10-03T02...  -76.943553  38.982470   
4  [\n  "2020-10-03T03:19:40Z",\n  "2020-10-03T03...  -76.938116  38.982084   

    END LONG    END LAT  
0 -76.934285  38.990739  
1 -76.949165  38.992849  
2 -76.933335  38.992676  
3 -76.943546  38.982476  
4 -76.943257  38.986101

In [36]:
original_dataset.shape

(40325, 16)

The dataset has 40325 rows and 16 columns

### Transforming the imported data to match the appropriate type

In [37]:
# Copy of imported data
new_dataset = original_dataset.copy()

In [38]:

# Convert as Timestamp data
new_dataset["CREATED"] = pd.to_datetime(new_dataset["CREATED"])
# Convert as Timestamp data
new_dataset["START"] = pd.to_datetime(new_dataset["START"])
# Convert as Timestamp data
new_dataset["END"] = pd.to_datetime(new_dataset["END"])
# Convert timestamp as list
new_dataset["TIMESTAMPS"] = new_dataset["TIMESTAMPS"].apply(eval)
# Get the path data as JSON
new_dataset["PATH"] = new_dataset["PATH"].apply(json.loads)
# Convert the path data to a list
new_dataset["PATH"] = new_dataset["PATH"].to_list()

In [45]:
# Verifying data types
new_dataset.dtypes

Unnamed: 0                    int64
Unnamed: 0.1                  int64
CREATED         datetime64[ns, UTC]
START           datetime64[ns, UTC]
END             datetime64[ns, UTC]
DISTANCE                    float64
RIDE_ID                       int64
CUSTOMER_ID                   int64
MINUTES                       int64
VEHICLE_TYPE                 object
PATH                         object
TIMESTAMPS                   object
START LONG                  float64
START LAT                   float64
END LONG                    float64
END LAT                     float64
dtype: object

In [40]:
new_dataset.head(5)

Unnamed: 0  Unnamed: 0.1                   CREATED  \
0           0             0 2020-10-03 03:32:15+00:00   
1           1             1 2020-10-03 03:07:39+00:00   
2           2             2 2020-10-03 03:17:05+00:00   
3           3             3 2020-10-03 02:24:58+00:00   
4           4             4 2020-10-03 03:19:34+00:00   

                      START                       END  DISTANCE  RIDE_ID  \
0 2020-10-03 03:32:15+00:00 2020-10-03 03:32:46+00:00     0.008  3135665   
1 2020-10-03 03:07:39+00:00 2020-10-03 03:37:06+00:00     1.062  3135395   
2 2020-10-03 03:17:05+00:00 2020-10-03 03:24:24+00:00     0.847  3135503   
3 2020-10-03 02:24:58+00:00 2020-10-03 03:11:56+00:00     3.447  3134955   
4 2020-10-03 03:19:34+00:00 2020-10-03 03:22:45+00:00     0.386  3135529   

   CUSTOMER_ID  MINUTES VEHICLE_TYPE  \
0       193964        1      Scooter   
1       542947       30      Scooter   
2       517321        8      Scooter   
3       578051       47      Scooter   
4       552305        4      Scooter   

                                                PATH  \
0  {'coordinates': [[-76.934163, 38.9906555], [-7...   
1  {'coordinates': [[-76.93818083333333, 38.98208...   
2  {'coordinates': [[-76.93759983333334, 38.97961...   
3  {'coordinates': [[-76.94355316666666, 38.98247...   
4  {'coordinates': [[-76.93811566666666, 38.98208...   

                                          TIMESTAMPS  START LONG  START LAT  \
0       [2020-10-03T03:32:14Z, 2020-10-03T03:32:30Z]  -76.934163  38.990656   
1  [2020-10-03T03:07:46Z, 2020-10-03T03:12:16Z, 2...  -76.938181  38.982087   
2  [2020-10-03T03:17:12Z, 2020-10-03T03:17:32Z, 2...  -76.937600  38.979613   
3  [2020-10-03T02:25:04Z, 2020-10-03T02:26:24Z, 2...  -76.943553  38.982470   
4  [2020-10-03T03:19:40Z, 2020-10-03T03:19:45Z, 2...  -76.938116  38.982084   

    END LONG    END LAT  
0 -76.934285  38.990739  
1 -76.949165  38.992849  
2 -76.933335  38.992676  
3 -76.943546  38.982476  
4 -76.943257  38.986101

In [44]:
# Sample path as list
new_dataset["PATH"][0]["coordinates"]

[[-76.934163, 38.9906555], [-76.93428483333334, 38.990738666666665]]

## Checking for missing values, Data processing and cleaning

In [14]:
# Checking for missing values
new_dataset.isna().any()

Unnamed: 0      False
Unnamed: 0.1    False
CREATED         False
START           False
END             False
DISTANCE         True
RIDE_ID         False
CUSTOMER_ID     False
MINUTES         False
VEHICLE_TYPE    False
PATH            False
TIMESTAMPS      False
START LONG      False
START LAT       False
END LONG        False
END LAT         False
dtype: bool

The distance column has missing values

In [46]:
new_dataset["DISTANCE"].isna().sum()
#dataset["DISTANCE"].dropna()

2273

Total of 2273 missing datapoints for the distance variables. No sure how to deal with this yet.

Next we separate all ride by years: October 2020 and October 2019

In [49]:
# Initialize new columns
new_dataset["Year"] = new_dataset["CREATED"][0].year

In [50]:
# Get the year of each data entry
for index, item in new_dataset.iterrows():
    new_dataset["Year"][index] = new_dataset["CREATED"][index].year

<ipython-input-50-ede04ab104ba>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_dataset["Year"][index] = new_dataset["CREATED"][index].year


In [51]:
ride_count_year = new_dataset.groupby("Year").size().reset_index(name='ride_count')
ride_count_year

Year  ride_count
0  2019       18382
1  2020       21943

In [54]:
new_dataset.head(3)

Unnamed: 0  Unnamed: 0.1                   CREATED  \
0           0             0 2020-10-03 03:32:15+00:00   
1           1             1 2020-10-03 03:07:39+00:00   
2           2             2 2020-10-03 03:17:05+00:00   

                      START                       END  DISTANCE  RIDE_ID  \
0 2020-10-03 03:32:15+00:00 2020-10-03 03:32:46+00:00     0.008  3135665   
1 2020-10-03 03:07:39+00:00 2020-10-03 03:37:06+00:00     1.062  3135395   
2 2020-10-03 03:17:05+00:00 2020-10-03 03:24:24+00:00     0.847  3135503   

   CUSTOMER_ID  MINUTES VEHICLE_TYPE  \
0       193964        1      Scooter   
1       542947       30      Scooter   
2       517321        8      Scooter   

                                                PATH  \
0  {'coordinates': [[-76.934163, 38.9906555], [-7...   
1  {'coordinates': [[-76.93818083333333, 38.98208...   
2  {'coordinates': [[-76.93759983333334, 38.97961...   

                                          TIMESTAMPS  START LONG  START LAT  \
0       [2020-10-03T03:32:14Z, 2020-10-03T03:32:30Z]  -76.934163  38.990656   
1  [2020-10-03T03:07:46Z, 2020-10-03T03:12:16Z, 2...  -76.938181  38.982087   
2  [2020-10-03T03:17:12Z, 2020-10-03T03:17:32Z, 2...  -76.937600  38.979613   

    END LONG    END LAT  Year  
0 -76.934285  38.990739  2020  
1 -76.949165  38.992849  2020  
2 -76.933335  38.992676  2020

# Separate dataset by years

In [57]:
# 2019 dataset
new_dataset_2019 = new_dataset.groupby(new_dataset.Year).get_group(2019)
# 2020 dataset
new_dataset_2020 = new_dataset.groupby(new_dataset.Year).get_group(2020)

In [19]:
import folium
from folium.plugins import MarkerCluster

In [33]:
# Initialize the global map using foolium
latitude = 0.0
longitude = 150
cp_longitude = -76.937144
cp_latitude = 38.989643
umd_latitude = 39.0457549
umd_longitude = -76.6412712


In [48]:
import folium
# construct a map centered at the Eiffel 
college_park = folium.Map(location = [umd_latitude, umd_longitude], zoom_start = 10)
# display the 
display(college_park)


In [27]:
paths_list

coordinates        type
0      [[-76.934163, 38.9906555], [-76.93428483333334...  MultiPoint
1      [[-76.93818083333333, 38.982087166666666], [-7...  MultiPoint
2      [[-76.93759983333334, 38.97961266666667], [-76...  MultiPoint
3      [[-76.94355316666666, 38.982470166666666], [-7...  MultiPoint
4      [[-76.93811566666666, 38.9820835], [-76.938261...  MultiPoint
...                                                  ...         ...
40320  [[-76.9439365, 38.98811], [-76.94407783333334,...  MultiPoint
40321  [[-76.93591716666667, 38.98638866666667], [-76...  MultiPoint
40322  [[-76.940231, 38.99637666666667], [-76.940162,...  MultiPoint
40323  [[-76.9441525, 38.991955833333336], [-76.94412...  MultiPoint
40324  [[-76.93416666666667, 38.99064533333333], [-76...  MultiPoint

[40325 rows x 2 columns]

In [33]:
paths_list["coordinates"][0]

[[-76.934163, 38.9906555], [-76.93428483333334, 38.990738666666665]]